02/app.js

```diff
@@ -14,11 +14,32 @@ class App extends React.Component {
 
         return (
             <section>
-                <Form />
+                <Form onClick={this.onClick} ref={ref => this.formRef = ref}/>
                 <List items={ usersList } />
             </section>
         )
     }
+
+    onClick = e => {
+        e.preventDefault();
+        const user = this.formRef.input.value;
+        if(user) {
+            this.setState({
+                usersList: [...this.state.usersList, user],
+            }, () => {
+                // Dlaczego to działa?
+                this.formRef.input.value = ''
+
+                //A to nie?
+                // user = '';
```

Comment:
Wcześniej napisałeś `const user = this.formRef.input.value;` czyli "skopiuj" wartość prymitywną (string) do `user`.
Napisanie `user = ''` (gdyby było to nie była stała) to powoduje zmianę tylko wartości dla zmiennej `user` napisanie `this.formRef.input.value` powoduje modyfikację drzewa DOM. Tutaj nie ma referencji, która mogła by zrobić tak jak sądziłeś, że będzie działać :)